### PR TITLE
Updated pytest version from 4.0.2 to 4.6.3.

### PR DESCRIPTION
**Pull request by devendra104**
Pytest version updated from 4.0.2 to 4.6.3

And resolved  pytest_namespace deprecation with pytest_configure.
To resolve this problem we created NestedDict class in a conftest.py file that creates a special dictionary which will be accessible via dotted notation.

Till now CLI test cases executed successfully with the latest pytest version.

```
 pytest -v tests/foreman/cli/
2019-06-12 15:53:44 - conftest - DEBUG - Registering custom pytest_configure=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.6, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /home/desingh/Robottelo_outdated_requirements/robottelo/env_set/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/desingh/Robottelo_outdated_requirements/robottelo
plugins: xdist-1.28.0, mock-1.10.4, cov-2.7.1, forked-1.0.2, services-1.3.1
collecting 1404 items                                                                                                                                                                                             2019-06-12 15:53:48 - conftest - DEBUG - BZ deselect is disabled in settingscollected 1463 items                                                                                                                                                                                              tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_create_report SKIPPED                                                                                                                           [  0%]
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_create_reports SKIPPED                                                                                                                          [  0%]
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_identify_hostname SKIPPED                                                                                                                       [  0%]
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_search_report SKIPPED                                                                                                                           [  0%]
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_update_timer SKIPPED                                                                                                                            [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_copy_with_same_name PASSED                                                                                                    [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_create_with_invalid_name PASSED                                                                                               [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_create_with_usage_limit_with_invalid_integers PASSED                                                                          [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_create_with_usage_limit_with_not_integers PASSED                                                                              [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_update_autoattach PASSED                                                                                                      [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_update_name PASSED                                                                                                            [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_update_usage_limit PASSED                                                                                                     [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_add_custom_product PASSED                                                                                                     [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_add_redhat_and_custom_products SKIPPED                                                                                        [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_add_redhat_product SKIPPED                                                                                                    [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_add_subscription_by_id SKIPPED                                                                                                [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_content_override PASSED                                                                                                       [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_copy_by_parent_id PASSED                                                                                                      [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_copy_by_parent_name PASSED                                                                                                    [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_copy_subscription SKIPPED                                                                                                     [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_content_and_check_enabled PASSED                                                                                       [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_using_old_name PASSED                                                                                                  [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_cv PASSED                                                                                                         [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_default_lce_by_id PASSED                                                                                          [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_default_lce_by_name PASSED                                                                                        [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_description PASSED                                                                                                [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_name PASSED                                                                                                       [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_non_default_lce PASSED                                                                                            [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_usage_limit_default PASSED                                                                                        [  1%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_create_with_usage_limit_finite PASSED                                                                                         [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_by_name PASSED                                                                                                         [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_by_org_label PASSED                                                                                                    [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_by_org_name PASSED                                                                                                     [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_manifest SKIPPED                                                                                                       [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_subscription SKIPPED                                                                                                   [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_with_cv PASSED                                                                                                         [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_delete_with_lce PASSED                                                                                                        [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_list_by_cv_id PASSED                                                                                                          [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_list_by_name PASSED                                                                                                           [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_remove_host_collection_by_id PASSED                                                                                           [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_remove_host_collection_by_name PASSED                                                                                         [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_remove_user PASSED                                                                                                            [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_aks_to_chost SKIPPED                                                                                                   [  2%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_aks_to_chost_in_one_command SKIPPED                                                                                    [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_autoattach PASSED                                                                                                      [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_autoattach_toggle PASSED                                                                                               [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_cv PASSED                                                                                                              [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_description PASSED                                                                                                     [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_host_collection PASSED                                                                                                 [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_host_collection_with_default_org PASSED                                                                                [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_lce PASSED                                                                                                             [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_name_by_id PASSED                                                                                                      [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_name_by_name PASSED                                                                                                    [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_usage_limit_to_finite_number PASSED                                                                                    [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_update_usage_limit_to_unlimited PASSED                                                                                        [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_usage_limit SKIPPED                                                                                                           [  3%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_positive_view_subscriptions_by_non_admin_user FAILED                                                                                   [  3%]
tests/foreman/cli/test_architecture.py::ArchitectureTestCase::test_negative_create_with_name PASSED                                                                                                         [  3%]
tests/foreman/cli/test_architecture.py::ArchitectureTestCase::test_negative_delete_by_id PASSED                                                                                                             [  4%]
tests/foreman/cli/test_architecture.py::ArchitectureTestCase::test_negative_update_name PASSED                                                                                                              [  4%]
tests/foreman/cli/test_architecture.py::ArchitectureTestCase::test_positive_create_with_name PASSED                                                                                                         [  4%]
tests/foreman/cli/test_architecture.py::ArchitectureTestCase::test_positive_delete_by_id PASSED                                                                                                             [  4%]
tests/foreman/cli/test_architecture.py::ArchitectureTestCase::test_positive_update_name PASSED                                                                                                              [  4%]
tests/foreman/cli/test_auth.py::HammerAuthTestCase::test_negative_no_credentials PASSED                                                                                                                     [  4%]
tests/foreman/cli/test_auth.py::HammerAuthTestCase::test_negative_no_permissions PASSED                                                                                                                     [  4%]
tests/foreman/cli/test_auth.py::HammerAuthTestCase::test_positive_change_session PASSED                                                                                                                     [  4%]
tests/foreman/cli/test_auth.py::HammerAuthTestCase::test_positive_create_session PASSED                                                                                                                     [  4%]
tests/foreman/cli/test_auth.py::HammerAuthTestCase::test_positive_disable_session PASSED                                                                                                                    [  4%]
tests/foreman/cli/test_auth.py::HammerAuthTestCase::test_positive_log_out_from_session PASSED                                                                                                               [  4%]
....
.....
....
```

### CODE REVIEW

**Comment by vijay8451** [comment]
@devendra104 Travis failed ,please check.

**Comment by san7ket** [comment]
NACK, needs extensive testing. Travis will fail as the new pytest as the changed the code of decorators. 
We need to look as what was the change and how it affects us from running decorated tests.

**Comment by devendra104** [comment]
!!!!Update!!!!!
The unitTest get broken because of "hasattr" method, hasattr method return True in version 4.0.2 but after upgrading in any of the version(4.1.0,4.6.2,4.6.3 etc) it always returns False.

`self.assertFalse(hasattr(foo, 'run_in_one_thread'))`

**Comment by mirzal** [comment]
On 4.0.2:
```
(Pdb) vars(foo)
{'bugzilla_ids': ['123456'], 'run_in_one_thread': <MarkInfo Mark(name='run_in_one_thread', args=(), kwargs={})>, 'pytestmark': [Mark(name='run_in_one_thread', args=(), kwargs={})]}
```

On 4.6.3:
```
(Pdb) vars(foo)
{'bugzilla_ids': ['123456'], 'pytestmark': [Mark(name='run_in_one_thread', args=(), kwargs={})]}
```

It's important to note that `run_in_one_thread` is just a marker (there is no real logic on Robottelo side, it's only used to select tests for CI runs - https://github.com/SatelliteQE/robottelo-ci/blob/master/scripts/satellite6-automation.sh#L103 ), and as long as pytest correctly recognizes it, it doesn't matter how it is stored internally. To verify that, I ran following command on 4.0.2 and 4.6.3:

```
pytest --collect-only -m 'tier1 and run_in_one_thread and not stubbed' tests/foreman/api/
```
They both find, select and deselect the same number of tests. It would be useful to re-run it and check results for other tiers and directories.

I would say failures that we are observing show problems in our tests, not in pytest.

As for `PytestUnknownMarkWarning` that Travis complains about - these are just warnings and can be ignored for time being. pytest would like us to register markers that we use, but it isn't mandatory. See https://docs.pytest.org/en/latest/mark.html for details how this can be done.

**Comment by mirzal** [comment]
And one more thing: this failure is related to `run_in_one_thread_if_bug_open` decorator, which we do not use in Foreman tests at all. There was only one file where we ever used it: 7c1df198e5e0adcd646182c4a8d3d5854f846cc2

**Comment by devendra104** [comment]
Pytest::Structure:: MarkDecorator class get changed in pytest 4.6.2, this change was effective in all upcoming version after 4.0.2.
One of the methods "store_legacy_markinfo" has removed from MarkDecorator class that creates an issue to add mark object in call function.

4.0.2--> _pytest->Mark->structures.py
def __call__(self, *args, **kwargs):
        if args and not kwargs:
            func = args[0]
            is_class = inspect.isclass(func)
            if len(args) == 1 and (istestfunc(func) or is_class):
                if is_class:
                    store_mark(func, self.mark)
                else:
                    store_legacy_markinfo(func, self.mark) --->This method added mark decorator as a func parameter and that parameter will be used by unittest(test_decorators.py::RunInOneThreadIfBugOpen::test_mark_one_thread_bugzilla_bug) for validation.
                    store_mark(func, self.mark)
                return func
        return self.with_args(*args, **kwargs)

All version after 4.0.2
def __call__(self, *args, **kwargs):

        if args and not kwargs:
            func = args[0]
            is_class = inspect.isclass(func)
            if len(args) == 1 and (istestfunc(func) or is_class):
                store_mark(func, self.mark)
                return func
        return self.with_args(*args, **kwargs)


Kindly suggest what would be the next optimized step to solve this issue because most of our test cases are using a run_in_one_thread decorator.
The pytest_namepsace(deprecated after all pytest 4.0 release) changes was working fine with 4.0.2 version.

**Comment by mirzal** [comment]
> Kindly suggest what would be the next optimized step to solve this issue because most of our test cases are using a run_in_one_thread decorator

If selection process results are unchanged in newer version of pytest, the way markers are stored internally is not our concern.

Our test checks if function object has specific property added by pytest. But this is spurious activity. If pytest decide to store marker data some other way, then our test will fail; but if pytest observable behavior didn't change, then there is no bug to fix. It's just test that must be update to check for what we really care about in correct way, instead of taking shortcuts.

If it was my choice, I would do these things:
- run `pytest --collect-only` for each tier and directory using pytest 4.0.2 and 4.6.3. This simulates what is happening on Jenkins. If both pytest versions find and deselect the same number of tests, we are good.
- remove `run_in_one_thread_if_bug_open` decorator completely and create issue to add it back. This is safe to do at the moment, as nothing uses this decorator (and, historically, hardly anyone used it).
- remove tests for decorator that is not here anymore

File: robottelo/decorators/__init__.py [hunk 1]
```diff
@@ -325,6 +324,7 @@ def __call__(self, func):
                 config_picker=self.config_picker
         )) or (self.bug_type == 'redmine' and rm_bug_is_open(self.bug_id)):
             func = self._wrapper(func)
+            func.run_in_one_thread = func.pytestmark
```
Comment (mirzal) [inline robottelo/decorators/__init__.py]: You are adding random property to function object just to make broken test happy. I strongly feel this is not the correct way to solve the issue.

**Comment by devendra104** [comment]
> * pytest --collect-only

Thanks @mirzal for help, Agree with your statement "If pytest decide to store marker data some other way, then our test will fail; but if pytest observable behavior didn't change, then there is no bug to fix."
Right rather than choosing the shortcut we have to update our unittest case accordingly.
 
I will follow the instruction that you suggested and will update you soon.

File: robottelo/decorators/__init__.py (hunk 1 above)
Comment (devendra104) [reply robottelo/decorators/__init__.py]: Agree, This is not a good practice, I will follow the instruction that you suggested in your previous comment and will update you soon.

**Comment by devendra104** [comment]
@mirzal I have checked the same for the rest of the tier across pytest:4.0.2 and 4.6.3 and find pytest deselect the same number of tests.
```
Tier1:
pytest:4.0.2 
===== 1003 deselected in 2.06 seconds =====
pytest:4.6.3
===== 1003 deselected in 1.97 seconds =====

Tier2:
pytest:4.0.2
====== 1034 deselected in 1.98 seconds =====
pytest: 4.6.3
===== 1034 deselected, 1 warnings in 2.15 seconds ====

Tier3:
pytest:4.0.2
===== 1041 deselected in 2.04 seconds ======
pytest4.6.3
===== 1041 deselected, 1 warnings in 2.15 seconds =====

Tier4:
pytest:4.0.2
===== 1046 deselected in 2.03 seconds =====
pytest:4.6.3
==== 1046 deselected, 1 warnings in 2.14 seconds ====

destructive:
pytest:4.0.2
====== 1058 deselected in 1.98 seconds =====
pytest:4.6.3
===== 1058 deselected, 1 warnings in 2.10 seconds ======

upgrade:
pytest:4.0.2
===== 1037 deselected in 2.03 seconds =====
pytest4:6.3
===== 1037 deselected, 1 warnings in 2.11 seconds =====

```
And find one common warning for the unknown mark  "PytestUnknownMarkWarning: Unknown pytest.mark.stubbed" 
So I think we are good to go with your suggestion to remove run_in_one_thread_if_bug_open decorator and create a separate issue to add it back.

**Comment by devendra104** [comment]
Test Result after upgrade from 4.0.2 to 4.6.3

```
BUG - Registering custom pytest_configure
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.6, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /home/desingh/Robottelo_outdated_requirements_PR/robottelo/pyenv/bin/python3.6
cachedir: .pytest_cache
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_identify_hostname SKIPPED                                                                                                                       [  0%]
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_search_report SKIPPED                                                                                                                           [  0%]
tests/foreman/cli/test_abrt.py::AbrtTestCase::test_positive_update_timer SKIPPED                                                                                                                            [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_copy_with_same_name PASSED                                                                                                    [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_create_with_invalid_name PASSED                                                                                               [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_create_with_usage_limit_with_invalid_integers PASSED                                                                          [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_create_with_usage_limit_with_not_integers PASSED                                                                              [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_update_autoattach PASSED                                                                                                      [  0%]
tests/foreman/cli/test_activationkey.py::ActivationKeyTestCase::test_negative_update_name PASSED                                                                                                            [  0%]
```

**Comment by devendra104** [comment]
Executed Tier1 test cases in which 84 test cases were failed out of 1074, And those failures were not related to pytest.
====== 60 failed, 989 passed, 146 warnings, 25 error in 10732.98 seconds =======

File: robottelo/decorators/__init__.py [hunk 2]
```diff
@@ -295,8 +295,10 @@ def config_picker():
 class run_in_one_thread_if_bug_open(robozilla_skip_if_bug_is_open, object):
     """A decorator that sets pytest marker and allows to select test that
     should be run sequentially only if bug is open.
+    TODO// Will be handled in new issue
```
Comment (mirzal) [inline robottelo/decorators/__init__.py]: We need better comment here. Something like:

> Removed as part of GitHub PR#7027, see Issue #7048

Comment (devendra104) [reply robottelo/decorators/__init__.py]: Same has been acknowledged.

**Pull request merged**